solving equations with Abs in solveset

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->

closes #10751 as a continuation of that work
closes #10532
closes #14485 
closes #14501 

#### Brief description of what is fixed or changed

see #10751

#### Other comments

Contains will no longer return relationals for things like `Contains(x, Interval(0, oo))`; it will now only evaluate if a Set is produced or evaluates to true or false. An alternative might be to make it `evaluate=False` by default and give Contains a doit method.

ConditionSet has an `_eval_subs` routine and `free_symbols` that tries to handle correctly the dummy symbol of the ConditionSet. It also allows the continued (ab)use of ConditionSet as a container for an unsolved system of equations.

Commit:
added _invert_abs function for inverting abs functions

Commit:
Added docstrings for the new _invert_abs function

Commit:
updated _invert_abs with new solveset API

Commit:
updated tests for _invert_abs

Commit:
added XFAILing solvify tests

Commit:
added XFAIL tests for Abs functions using nonlinsolve

smichr committed:
Contains -> Contains, Set, S.true or S.false

Don't allow Contains to be re-expressed as
a Boolean, e.g. Contains(n, Interval(0, oo)) no
longer returns as (n >= 0) & (n < oo). To get that
form, use the method:

>>> Interval(0,oo).contains(n)
(n >= 0) & (n < oo)

smichr commented:
ping @kshitij10496 
ping @Yathartha22

Yathartha22 commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -53,15 +54,19 @@ def test_invert_real():
+        x, Contains(n, Interval(0, oo), evaluate=False), {-n, n})
+
+    conditions = Contains(n, Interval(0, oo), evaluate=False)
+    assert solveset(abs(x) - n, x, S.Reals) == ConditionSet(
```
tests are repeated.

Yathartha22 replied:
Also `evaluate=False` should be removed now (as per the changes done in `Contains` class above).

smichr replied:
Will correct when I update PR.

Yathartha22 commented:
I will update you in few hours after checking it thoroughly.

smichr committed:
move abs test; delete dup

smichr commented:
Ability to solve this has been lost:
```
>>> solveset(Eq(sin(Abs(x)), 1), x, domain=S.Reals)
Union(
Intersection(Interval(0, oo), ImageSet(Lambda(_n, (-1)**_n*pi/2 + _n*pi), S.Integers)),
Intersection(Interval(-oo, 0), ImageSet(Lambda(_n, _n*pi - (-1)**(-_n)*pi/2), S.Integers)))
```
The error ends with
```
...
    elif len(condition_ys) == 1:
TypeError: object of type 'ImageSet' has no len()
```

Yathartha22 commented:
> Ability to solve this has been lost:

This is mainly because the `values` returned is `ImageSet(Lambda(_n, (-1)**_n*pi/2 + _n*pi), S.Integers)`, it gets into [second if condition](https://github.com/smichr/sympy/blob/bee672654e4e68b170a7bade7fa0eea6e3d44fb8/sympy/solvers/solveset.py#L362) where it encounters the error you mentioned.

Two things that can be done:

 - Either we check for cases when `values` is an `ImageSet`, or

 - add a check in the second if condition for `ImageSet`.

Both the fixes depends on the fact that what should be the answer for such cases.

Should it be something like this:?? (I am not sure about this result)

```
>>> solveset(Eq(sin(Abs(x)), 1), x, domain=S.Reals)
ConditionSet(x, Contains((Lambda(_n, (-1)**_n*pi/2 + _n*pi), S.Integers), Interval(0, oo)), Union(ImageSet(Lambda(_n, (-1)**_n*pi/2 + _n*pi), S.Integers), ImageSet(Lambda(_n, _n*pi - (-1)**(-_n)*pi/2), S.Integers)))
```

Yathartha22 commented on sympy/solvers/solveset.py:
```diff
@@ -314,6 +311,81 @@ def _invert_complex(f, g_ys, symbol):
+    g_x, values = _invert_real(f_n, Union(imageset(Lambda(n, n), g_ys),
+                               imageset(Lambda(n, -n), g_ys)), symbol)
+
+    if isinstance(values, ConditionSet):
```
@smichr I might be wrong but I don't understand this. Why we need to check for `ConditionSet`, `_invert_real` never returns a `ConditionSet`, hence this will never be True.

smichr replied:
I suspect it does but can check. The reason I think it does is because I had to watch out for the Contains using evaluate True or False (before making it so Contains no longer returns a Relational).

But I didn't write the code -- I just did the minimal amount to get #10751 up and running again.

smichr replied:
I pruned out some code. Let's see if tests pass.

smichr committed:
remove uncovered code

smichr committed:
modify test 14454

smichr committed:
denest ConditionSet

smichr committed:
raise error for non-FiniteSet

smichr commented:
I am not adept enough with sets to know how to handle the non-Finite set case. Rewriting as Piecewise and trying to solve fails, too, so I am leaving the XFAIL test.

Yathartha22 commented:
I will try to figure it out @smichr , otherwise you merge it as it is and we fix it later in another PR (may be as part of GSoC)

smichr commented:
> otherwise you merge it as it is and we fix it later

ok

smichr commented on sympy/sets/conditionset.py:
```diff
@@ -48,6 +48,11 @@ def __new__(cls, sym, condition, base_set):
             else:
                 return FiniteSet(*sifted[True])
+        if isinstance(base_set, ConditionSet):
+            s, c, b = base_set.args
```
Think about this some more with examples.

Yathartha22 commented on sympy/solvers/solveset.py:
```diff
@@ -314,6 +311,48 @@ def _invert_complex(f, g_ys, symbol):
+        else:
+            conditions = True
+    return g_x, ConditionSet(g_x, conditions, values)
+
```
Can we use the following, it might slight reduce computation

```diff

diff --git a/sympy/solvers/solveset.py b/sympy/solvers/solveset.py
index 08c6253..980bf3c 100644
--- a/sympy/solvers/solveset.py
+++ b/sympy/solvers/solveset.py
@@ -338,18 +338,13 @@ def _invert_abs(f, g_ys, symbol):
     if not isinstance(g_ys, FiniteSet):
         raise NotImplementedError
     else:
-        unknown = []
+        conditions = True
         for a in g_ys.args:
             ok = a.is_nonnegative if a.is_Number else a.is_positive
             if ok is None:
-                unknown.append(a)
+                conditions = And(*[Contains(a, Interval(0, oo))])
             elif not ok:
                 return g_x, S.EmptySet
-        if unknown:
-            conditions = And(*[Contains(i, Interval(0, oo))
-                for i in unknown])
-        else:
-            conditions = True
     return g_x, ConditionSet(g_x, conditions, values)

```

smichr replied:
I'm not sure that that will be faster. Do you have some experience that says it will? Typically, I try to avoid incremental updates to a return value (e.g. collect args then return `Add(*args)` instead of doing `rv += arg` in a loop).

Yathartha22 replied:
Since two loops need to be traversed, time complexity may slightly increase to `O(m+n)` (where m is len(g_ys.args), n is len(unknown)), whereas the above would lead to `O(m)`. This won't be of much difference.
I suggested because why to unnecessary traverse another loop when this can be achieved with one, and also it might reduce some code.

smichr replied:
@Yathartha22 , here is an example of what I mean:
```
>>> timeit('''
... args = []
... for i in arg:
...   args.append(i)
... And(*args)
... ''', '''from sympy import var, And\narg = var('a:z')''', number=100)
0.23920276681502628
>>> timeit('''
... rv = True
... for i in arg:
...   rv = And(rv, i)
... ''', '''from sympy import var, And\narg = var('a:z')''', number=100)
0.8529278468695338
```

Yathartha22 replied:
Yes @smichr I also checked with and interestingly it seems that the one I suggested would take longer. Any reason why the approach takes longer time even when updation takes place in parallel.

smichr committed:
reorganize _invert_abs

smichr committed:
Revert "denest ConditionSet"

This reverts commit 364c49e46aac966d220d5e092f4520203ab9352c.

More thought needs to go into what to do with

ConditionSet(x, x>1, ConditionSet(y, y>x, S.Integers))

Are those the same x?? Should the answer be

ConditionSet(y, And(y>1, y>x), S.Integers)

What if it were

ConditionSet(x, x>y, ConditionSet(y, y>2, S.Integers))

Should that be

ConditionSet(x, And(x > y, x > 2), S.Integers)

smichr committed:
modify ConditionSet

changes to instantiation involve more checking and standardization
and collapsing of nested ConditionSets

methods added for
* free_symbols
* _eval_subs
* dummy_eq

All lines are tested.

jksuom commented on sympy/sets/conditionset.py:
```diff
@@ -33,6 +34,13 @@ class ConditionSet(Set):
+        if not isinstance(sym, Symbol):
+            raise TypeError('sym must be a symbol')
+        condition = as_Boolean(condition)
+        if isinstance(base_set, set):
```
I would check that `sym` does not appear in `base_set`. Only `condition` may depend on it.

smichr replied:
That check happens below at line 77 after all the sym-wrangling is finished.

smichr commented:
@jksuom , this seems like an abuse of ConditionSet.

```
def test_nonlinsolve_conditionset():
    # when solveset failed to solve all the eq
    # return conditionset
    f = Function('f')
    f1 = f(x) - pi/2
    f2 = f(x) - 3*pi/2
    intermediate_system = FiniteSet(2*f(x) - 3*pi, 2*f(x) - pi)
    symbols = FiniteSet(x, y)
    soln = ConditionSet(
        symbols,
        intermediate_system,
        S.Complexes)
    assert nonlinsolve([f1, f2], [x,y]) == soln
```

BTW, I think f2 should be `f(y) - 3*pi/2`. The system here should have returned an empty set.

I don't think we have an object to represent an unsolved system of equations.

jksuom commented:
It really looks like there is a typo in `f2`. If it were `f(y) - 3*pi/2`, the condition set could make sense given that `f` is unknown. With `x` in `f2` there are clearly no solutions (and `soln` should be `EmptySet`, I think).

smichr committed:
concession for unsolved equations

This can be reverted once a new syntax for an unsolved equation
set is introduced.

smichr commented:
OK, what to do with ImageSet in `_invert_abs` still remains and how best to represent unsolved sets of equations. I will open two new issues for those. I will commit this in about 12 hours if there is no more comment.

smichr committed:
_invert_abs handles non-FiniteSet

smichr commented:
@Yathartha22 , the previous solution was right in the OP of #10532 but just complicated (for FiniteSets). So now the non-FiniteSet case is the same (so the `sin(abs(x))` still resolves) but the FiniteSet cases are returned as a ConditionSet.

Yathartha22 commented:
As of now LGTM +1

smichr commented:
OK, thanks for the help with this @Yathartha22 . It's in.
